### PR TITLE
fix(youtube-video): keep current time updating during playback

### PR DESCRIPTION
**Pull request by mihar-22**
## Summary

Keep YouTube playback time and the player controls updating when the iframe API reports clock changes in coarse intervals. Playback progress is no longer mistaken for a seek while the YouTube player is actively playing.

## Changes

- Ignore seek inference from current-time jumps while the YouTube player reports `PLAYING`
- Cover coarse playback updates when buffered progress is not yet available

## Testing

- `pnpm -F @videojs/youtube-video test`
- `pnpm -F @videojs/youtube-video exec vp pack`
- `pnpm typecheck`
- Manual React sandbox verification: current-time text and seek slider advance during YouTube playback

---

> [!NOTE]
> **Medium Risk**
> Changes seek/timeupdate polling semantics in the YouTube media adapter, which affects all consumers of seeking and currentTime during playback.
> 
> **Overview**
> Fixes **YouTube adapter** playback UI stalling when the iframe API reports **coarse** `getCurrentTime()` steps: polling no longer treats those jumps as user seeks while the player is **PLAYING**, so **`timeupdate`** and controls keep advancing.
> 
> **Programmatic seeks** (`currentTime = …`) now track **target**, **origin**, and a **settle timeout**, emit **`seeking`** once, call **`seekTo`**, hold the reported time at the requested value until the clock is within tolerance (or the settle window passes for no-op/nearby seeks), then emit **`seeked`**. Overlapping seeks wait for the **latest** target; **`PLAYING`** state changes only clear seeking when there is no active seek target.
> 
> Adds focused unit tests for coarse playback, forward/back seeks, overlapping seeks, and nearby/no-op seek edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b64b386a936427d3e0cf7a7d07b162049bd407fb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by sampotts — APPROVED** [review]

File: packages/adapters/youtube-video/src/adapter.ts [hunk 1]
```diff
@@ -655,7 +655,11 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
     if (this.#seeking && bufferedEnd > 0.1) {
       this.#seeking = false;
       this.dispatchEvent(new Event('seeked'));
-    } else if (!this.#seeking && Math.abs(time - this.#currentTime) > 0.1) {
+    } else if (
+      !this.#seeking &&
+      player.getPlayerState() !== STATE_PLAYING &&
+      Math.abs(time - this.#currentTime) > 0.1
+    ) {
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts]: Good catch. Addressed in 98927d66b by explicitly entering the adapter seeking state when a programmatic `currentTime` update reaches `seekTo`. Added coverage for a mid-playback seek completing and subsequent coarse `timeupdate` events continuing normally.

File: packages/adapters/youtube-video/src/adapter.ts [hunk 2]
```diff
@@ -250,7 +250,14 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
 
     this.#currentTime = value;
     // `seekTo` keeps the player paused when called from a paused state.
-    this.#afterLoad((p) => p.seekTo(value, true));
+    this.#afterLoad((p) => {
+      if (!this.#seeking) {
+        this.#seeking = true;
+        this.dispatchEvent(new Event('seeking'));
+      }
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts:267]: Fixed in 7f794ac06. Programmatic seeks now retain their requested target and only emit `seeked` once the polled YouTube clock reaches that target. Until then, stale poll values are suppressed so the optimistic current time cannot snap back. The regression test now covers both a nonzero buffered fraction and a repeated `PLAYING` event before the iframe clock moves.

File: packages/adapters/youtube-video/src/adapter.ts [hunk 3]
```diff
@@ -651,20 +662,26 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
...
+    if ((this.#seekTarget === null || reachedSeekTarget) && time !== this.#currentTime) {
       this.#currentTime = time;
       this.dispatchEvent(new Event('timeupdate'));
     }
 
+    if (wasSeeking && (reachedSeekTarget || (this.#seekTarget === null && bufferedEnd > 0.1))) {
+      this.#seeking = false;
+      this.#seekTarget = null;
+      this.dispatchEvent(new Event('seeked'));
+    }
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts:706]: Fixed in 8b5877dba. The adapter now records the YouTube clock at the start of a programmatic seek and treats the seek as complete once the clock reaches or crosses the requested target in that direction, rather than requiring an exact ±0.1s sample. Added regression coverage for coarse forward and backward post-seek samples.

File: packages/adapters/youtube-video/src/adapter.ts [hunk 4]
```diff
@@ -651,20 +665,30 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
     const time = player.getCurrentTime();
     const duration = player.getDuration();
     const bufferedEnd = player.getVideoLoadedFraction() * duration;
+    const wasSeeking = this.#seeking;
+    const reachedSeekTarget =
+      this.#seekTarget !== null &&
+      this.#seekOrigin !== null &&
+      (this.#seekTarget >= this.#seekOrigin ? time >= this.#seekTarget - 0.1 : time <= this.#seekTarget + 0.1);
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts]: Fixed in f2f22dacd. Seek completion now also accepts a substantial jump toward the requested position, covering YouTube landing on a nearby keyframe on either side without accepting the unchanged pre-seek clock. The forward and backward tests now use samples just short of their targets.

File: packages/adapters/youtube-video/src/adapter.ts [hunk 5]
```diff
@@ -644,27 +658,46 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
...
+  #hasAppliedSeek(time: number) {
+    const target = this.#seekTarget;
+    const origin = this.#seekOrigin;
+    if (target === null || origin === null) return false;
+
+    const crossedTarget = target >= origin ? time >= target - 0.1 : time <= target + 0.1;
+    // YouTube can land on a nearby keyframe without crossing the requested time.
+    const jumpedTowardTarget = Math.abs(time - target) <= Math.abs(origin - target) / 2;
+
+    return crossedTarget || jumpedTowardTarget;
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts]: Fixed in ebbde0748. The adapter now tracks targets superseded by rapid follow-up seeks and rejects clock samples that still belong to an older target. Added a double-seek regression test that verifies the first landing keeps the latest optimistic time and seeking state until the latest target is applied.

File: packages/adapters/youtube-video/src/adapter.ts [hunk 6]
```diff
@@ -644,27 +668,52 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
...
 
+  #hasAppliedSeek(time: number) {
+    const target = this.#seekTarget;
+    const origin = this.#seekOrigin;
+    if (target === null || origin === null) return false;
+
+    const belongsToSupersededSeek = this.#supersededSeekTargets.some(
+      (supersededTarget) => Math.abs(time - supersededTarget) <= Math.abs(time - target)
+    );
+    if (belongsToSupersededSeek) return false;
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts]: Fixed in bd4be866e. Removed superseded-target history entirely. Programmatic seek completion now compares the YouTube clock only with the latest requested target using a one-second tolerance for coarse samples and nearby keyframes. This preserves rejection of distant stale/older landings without allowing slider-drag history to keep seeking stuck; all coarse, backward, and overlapping-seek regressions remain covered.

File: packages/adapters/youtube-video/src/adapter.ts [hunk 7]
```diff
@@ -644,27 +657,41 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
     }
   }
 
+  #hasAppliedSeek(time: number) {
+    const target = this.#seekTarget;
+    if (target === null) return false;
+
+    // YouTube can report coarse clock samples or land on a nearby keyframe.
+    return Math.abs(time - target) <= SEEK_TOLERANCE;
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts]: Fixed in 2465bcfe2. Programmatic seeks now retain a pre-seek clock snapshot and require the iframe clock to move before the latest-target tolerance can complete the seek. Added a nearby-seek regression proving an unchanged in-tolerance clock preserves the optimistic time and seeking state until YouTube moves.

File: packages/adapters/youtube-video/src/adapter.ts [hunk 8]
```diff
@@ -644,27 +660,43 @@ export class YouTubeAdapter extends MediaPlayedRangesMixin(EventTarget) implemen
     }
   }
 
+  #hasAppliedSeek(time: number) {
+    const target = this.#seekTarget;
+    const origin = this.#seekOrigin;
+    if (target === null || origin === null) return false;
+
+    // YouTube can report coarse clock samples or land on a nearby keyframe.
+    return time !== origin && Math.abs(time - target) <= SEEK_TOLERANCE;
```
[the remark this answers is not in the document]
Comment (mihar-22) [reply packages/adapters/youtube-video/src/adapter.ts]: Valid edge case. I added a one-second settle fallback for an unchanged clock that is still within the YouTube keyframe tolerance. Immediate polling still requires real clock movement, so a stale pre-seek sample cannot emit seeked early; after the settle window, a legitimate same-position or nearby-keyframe landing no longer leaves the adapter frozen in seeking. Added a regression test in b64b386a9.

**Review by luwes — APPROVED** [review]
thanks!
